File: app/lib/models/piece.dart
import 'dart:convert';
import 'dart:typed_data';

class Piece {
  final int id;
  final int userId;
  final String title;
  final String? discipline;
  final String? year;
  final String? edition;
  final int priceCents;
  final int? oldPriceCents;
  final int? costPriceCents;
  final bool rental;
  final int stock;
  final List<String>? shipsTo;
  final String? packaging;
  final String status;
  final List<PieceImageData> images;
  final DateTime createdAt;

  const Piece({
    required this.id,
    required this.userId,
    required this.title,
    this.discipline,
    this.year,
    this.edition,
    required this.priceCents,
    this.oldPriceCents,
    this.costPriceCents,
    required this.rental,
    required this.stock,
    this.shipsTo,
    this.packaging,
    required this.status,
    this.images = const [],
    required this.createdAt,
  });

  factory Piece.fromJson(Map<String, dynamic> json) => Piece(
        id: json['id'] as int,
        userId: json['user_id'] as int,
        title: json['title'] as String,
        discipline: json['discipline'] as String?,
        year: json['year'] as String?,
        edition: json['edition'] as String?,
        priceCents: json['price_cents'] as int,
        oldPriceCents: json['old_price_cents'] as int?,
        costPriceCents: json['cost_price_cents'] as int?,
        rental: json['rental'] as bool? ?? false,
        stock: json['stock'] as int? ?? 1,
        shipsTo: (json['ships_to'] as List?)?.cast<String>(),
        packaging: json['packaging'] as String?,
        status: json['status'] as String? ?? 'active',
        images: (json['images'] as List?)
                ?.map(
                    (i) => PieceImageData.fromJson(i as Map<String, dynamic>))
                .toList() ??
            [],
        createdAt: DateTime.parse(json['created_at'] as String),
      );

  String get priceFormatted {
    final euros = priceCents ~/ 100;
    return '\u20AC$euros';
  }
}

/// Lightweight model for list views (maps to PieceListOut).
class PieceListItem {
  final int id;
  final String title;
  final String? discipline;
  final String? year;
  final int priceCents;
  final bool rental;
  final String status;
  final Uint8List? coverImageBytes;
  final DateTime createdAt;

  const PieceListItem({
    required this.id,
    required this.title,
    this.discipline,
    this.year,
    required this.priceCents,
    required this.rental,
    required this.status,
    this.coverImageBytes,
    required this.createdAt,
  });

  factory PieceListItem.fromJson(Map<String, dynamic> json) => PieceListItem(
        id: json['id'] as int,
        title: json['title'] as String,
        discipline: json['discipline'] as String?,
        year: json['year'] as String?,
        priceCents: json['price_cents'] as int,
        rental: json['rental'] as bool? ?? false,
        status: json['status'] as String? ?? 'active',
        coverImageBytes: json['cover_image_b64'] != null
            ? base64Decode(json['cover_image_b64'] as String)
            : null,
        createdAt: DateTime.parse(json['created_at'] as String),
      );

  String get priceFormatted {
    final euros = priceCents ~/ 100;
    return '\u20AC$euros';
  }
}

class PieceImageData {
  final int id;
  final int position;
  final Uint8List bytes;

  const PieceImageData({
    required this.id,
    required this.position,
    required this.bytes,
  });

  factory PieceImageData.fromJson(Map<String, dynamic> json) => PieceImageData(
        id: json['id'] as int,
        position: json['position'] as int,
        bytes: base64Decode(json['image_b64'] as String),
      );
}

File: backend/app/schemas/piece.py
import base64
import json
from datetime import datetime

from pydantic import BaseModel

from ..models.piece import Piece, PieceImage


class PieceImageOut(BaseModel):
    id: int
    position: int
    image_b64: str

    model_config = {"from_attributes": True}

    @classmethod
    def from_model(cls, img: PieceImage) -> "PieceImageOut":
        return cls(
            id=img.id,
            position=img.position,
            image_b64=base64.b64encode(img.image_data).decode("ascii"),
        )


class PieceCreate(BaseModel):
    title: str
    discipline: str | None = None
    year: str | None = None
    edition: str | None = None
    description: str | None = None
    price_cents: int
    old_price_cents: int | None = None
    cost_price_cents: int | None = None
    rental: bool = False
    rental_daily_rate_cents: int | None = None
    stock: int = 1
    ships_to: list[str] | None = None
    packaging: str | None = None


class PieceUpdate(BaseModel):
    title: str | None = None
    discipline: str | None = None
    year: str | None = None
    edition: str | None = None
    description: str | None = None
    price_cents: int | None = None
    old_price_cents: int | None = None
    cost_price_cents: int | None = None
    rental: bool | None = None
    rental_daily_rate_cents: int | None = None
    stock: int | None = None
    ships_to: list[str] | None = None
    packaging: str | None = None


class PieceOut(BaseModel):
    id: int
    user_id: int
    title: str
    discipline: str | None = None
    year: str | None = None
    edition: str | None = None
    description: str | None = None
    price_cents: int
    old_price_cents: int | None = None
    cost_price_cents: int | None = None
    rental: bool
    rental_daily_rate_cents: int | None = None
    stock: int
    ships_to: list[str] | None = None
    packaging: str | None = None
    status: str
    images: list[PieceImageOut] = []
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

    @classmethod
    def from_model(cls, piece: Piece) -> "PieceOut":
        ships_to_list = None
        if piece.ships_to:
            ships_to_list = json.loads(piece.ships_to)

        return cls(
            id=piece.id,
            user_id=piece.user_id,
            title=piece.title,
            discipline=piece.discipline,
            year=piece.year,
            edition=piece.edition,
            description=piece.description,
            price_cents=piece.price_cents,
            old_price_cents=piece.old_price_cents,
            cost_price_cents=piece.cost_price_cents,
            rental=piece.rental,
            rental_daily_rate_cents=piece.rental_daily_rate_cents,
            stock=piece.stock,
            ships_to=ships_to_list,
            packaging=piece.packaging,
            status=piece.status,
            images=[PieceImageOut.from_model(img) for img in piece.images],
            created_at=piece.created_at,
            updated_at=piece.updated_at,
        )


class PieceListOut(BaseModel):
    """Lighter version without full image data for list endpoints."""

    id: int
    title: str
    discipline: str | None = None
    year: str | None = None
    price_cents: int
    rental: bool
    status: str
    cover_image_b64: str | None = None
    created_at: datetime

    @classmethod
    def from_model(cls, piece: Piece) -> "PieceListOut":
        cover_b64 = None
        if piece.images:
            first = min(piece.images, key=lambda img: img.position)
            cover_b64 = base64.b64encode(first.image_data).decode("ascii")

        return cls(
            id=piece.id,
            title=piece.title,
            discipline=piece.discipline,
            year=piece.year,
            price_cents=piece.price_cents,
            rental=piece.rental,
            status=piece.status,
            cover_image_b64=cover_b64,
            created_at=piece.created_at,
        )
